test_vm/src/lib.rs

```diff
@@ -606,6 +812,115 @@ pub fn actor(code: Cid, head: Cid, seq: u64, bal: TokenAmount) -> Actor {
     Actor { code, head, call_seq_num: seq, balance: bal }
 }
 
+pub struct InvocationTrace {
+    pub msg: InternalMessage,
+    pub code: Option<ExitCode>,
+    pub ret: Option<RawBytes>,
+    pub subinvocations: Vec<InvocationTrace>,
+}
+
+pub struct ExpectInvocation {
+    pub to: Address,       // required
+    pub method: MethodNum, // required
+    pub code: Option<ExitCode>,
+    pub from: Option<Address>,
+    pub params: Option<RawBytes>,
+    pub ret: Option<RawBytes>,
+    pub subinvocs: Option<Vec<ExpectInvocation>>,
+}
+
+impl ExpectInvocation {
+    // testing method that panics on no match
+    pub fn matches(&self, invoc: &InvocationTrace) {
+        let id = format!("[{}:{}]", invoc.msg.to, invoc.msg.method);
+        self.quick_match(invoc, "".to_string());
+        if let Some(c) = self.code {
+            assert_ne!(
+                None,
+                invoc.code,
+                "{} unexpected code: expected:{}was:{}",
+                id,
+                c,
+                ExitCode::OK
+            );
+            assert_eq!(
+                c,
+                invoc.code.unwrap(),
+                "{} unexpected code expected:{}was:{}",
+                id,
+                c,
+                invoc.code.unwrap()
+            );
+        }
+        if let Some(f) = self.from {
+            assert_eq!(
+                f, invoc.msg.from,
+                "{} unexpected from addr: expected:{}was:{} ",
+                id, f, invoc.msg.from
+            );
+        }
+        if let Some(p) = &self.params {
+            assert_eq!(
+                p, &invoc.msg.params,
+                "{} unexpected params: expected:{:x?}was:{:x?}",
+                id, p, invoc.msg.params
+            );
+        }
+        if let Some(r) = &self.ret {
+            assert_ne!(None, invoc.ret, "{} unexpected ret: expected:{:x?}was:None", id, r);
+            let ret = &invoc.ret.clone().unwrap();
+            assert_eq!(r, ret, "{} unexpected ret: expected:{:x?}was:{:x?}", id, r, ret);
+        }
+        if let Some(expect_subinvocs) = &self.subinvocs {
+            let subinvocs = &invoc.subinvocations;
+
+            let panic_str = format!(
+                "unexpected subinvocs:\n expected: \n[\n{}]\n was:\n[\n{}]\n",
+                self.fmt_expect_invocs(expect_subinvocs),
+                self.fmt_invocs(subinvocs)
+            );
+            assert!(subinvocs.len() == expect_subinvocs.len(), "{}", panic_str);
```

Comment:
```suggestion
            assert!(subinvocs.len() == expect_subinvocs.len(),
                    "unexpected subinvocs:\n expected: \n[\n{}]\n was:\n[\n{}]\n",
                    self.fmt_expect_invocs(expect_subinvocs),
                    self.fmt_invocs(subinvocs));
```

(lets one avoid doing any work if the assertion succeeds)

Reply:
This makes sense but we use the same string in the loop below so it's nice to assign for clarity.

Reply:
Ah, I see.